pkg/member/member.go

```diff
@@ -44,8 +44,59 @@ const (
 	dcLocationConfigEtcdPrefix = "dc-location"
 )
 
-// Member is used for the election related logic.
-type Member struct {
+// Member defines the interface for the election related logic.
+type Member interface {
+	// ID returns the unique ID for this participant in the group. For example, it can be
+	// unique server id of a cluster or the unique keyspace group replica id of the election
+	// group comprised of the replicas of a keyspace group.
+	ID() uint64
+	// MemberValue returns the member value.
+	MemberValue() string
+	// Member returns the member.
+	Member() *pdpb.Member
+	// Client returns the etcd client.
+	Client() *clientv3.Client
+	// IsLeader returns whether the participant is the leader or not by checking its
+	// leadership's lease and leader info.
+	IsLeader() bool
+	// GetLeaderID returns current leader's member ID.
+	GetLeaderID() uint64
+	// GetLeader returns current leader of the election group.
+	GetLeader() *pdpb.Member
+	// EnableLeader declares the member itself to be the leader.
+	EnableLeader()
+	// GetLeaderPath returns the path of the leader.
+	GetLeaderPath() string
+	// GetLeadership returns the leadership of the PD member.
+	GetLeadership() *election.Leadership
+	// CampaignLeader is used to campaign a PD member's leadership
+	// and make it become a PD leader.
+	CampaignLeader(leaseTimeout int64) error
+	// KeepLeader is used to keep the PD leader's leadership.
+	KeepLeader(ctx context.Context)
+	// CheckLeader checks returns true if it is needed to check later.
+	CheckLeader() (*pdpb.Member, int64, bool)
+	// WatchLeader is used to watch the changes of the leader.
+	WatchLeader(serverCtx context.Context, leader *pdpb.Member, revision int64)
+	// ResetLeader is used to reset the PD member's current leadership.
+	// Basically it will reset the leader lease and unset leader info.
+	ResetLeader()
+	// isSameLeader checks whether a server is the leader itself.
+	isSameLeader(leader *pdpb.Member) bool
+	// CheckPriority checks whether there is another participant has higher priority and resign it as the leader if so.
+	CheckPriority(ctx context.Context)
+	// GetDCLocationPathPrefix returns the dc-location path prefix of the cluster.
+	GetDCLocationPathPrefix() string
+	// GetDCLocationPath returns the dc-location path of a member with the given member ID.
+	GetDCLocationPath(id uint64) string
+	// CheckLeaderPrecheck does some pre-check before checking whether or not it's the leader.
```

Comment:
// CheckLeaderPrecheck does some pre-check before checking whether it's the leader.

Reply:
changed.